// @theia/preferences/lib/browser/views/components/index.js
"use strict";
/********************************************************************************
 * Copyright (C) 2020 Ericsson and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
var preference_array_input_1 = require("./preference-array-input");
Object.defineProperty(exports, "PreferenceArrayInput", { enumerable: true, get: function () { return preference_array_input_1.PreferenceArrayInput; } });
var preference_boolean_input_1 = require("./preference-boolean-input");
Object.defineProperty(exports, "PreferenceBooleanInput", { enumerable: true, get: function () { return preference_boolean_input_1.PreferenceBooleanInput; } });
var preference_json_input_1 = require("./preference-json-input");
Object.defineProperty(exports, "PreferenceJSONInput", { enumerable: true, get: function () { return preference_json_input_1.PreferenceJSONInput; } });
var preference_number_input_1 = require("./preference-number-input");
Object.defineProperty(exports, "PreferenceNumberInput", { enumerable: true, get: function () { return preference_number_input_1.PreferenceNumberInput; } });
var preference_select_input_1 = require("./preference-select-input");
Object.defineProperty(exports, "PreferenceSelectInput", { enumerable: true, get: function () { return preference_select_input_1.PreferenceSelectInput; } });
var preference_string_input_1 = require("./preference-string-input");
Object.defineProperty(exports, "PreferenceStringInput", { enumerable: true, get: function () { return preference_string_input_1.PreferenceStringInput; } });
